### PR TITLE
Revise nullability in HttpEntity and ResponseEntity

### PR DESCRIPTION
**Pull request by ngocnhan-tran1996**
Based on the IDE warning shown in the screenshot

image

This change adds `@Nullable` to the generic bound of ResponseEntity to make the nullability contract.

Related to gh-36236

### CODE REVIEW

File: spring-web/src/main/java/org/springframework/http/HttpEntity.java
Comment (sdeleuze) [inline spring-web/src/main/java/org/springframework/http/HttpEntity.java]: Here, you should just leverage the nullability defined at type level, so please remove the explicit `@Nullable` override here.

Comment (sdeleuze) [inline spring-web/src/main/java/org/springframework/http/HttpEntity.java]: Here, you should just leverage the nullability defined at type level, so please remove the explicit `@Nullable` override here and in other similar use cases.

File: spring-web/src/main/java/org/springframework/http/ResponseEntity.java [hunk 1]
```diff
@@ -261,7 +261,7 @@ public static BodyBuilder ok() {
 	 * @return the created {@code ResponseEntity}
 	 * @since 4.1
 	 */
-	public static <T> ResponseEntity<T> ok(@Nullable T body) {
+	public static <T extends @Nullable Object> ResponseEntity<T> ok(@Nullable T body) {
```
Comment (sdeleuze) [inline spring-web/src/main/java/org/springframework/http/ResponseEntity.java]: Here I think you should just use `public static <T> ResponseEntity<T> ok(T body)` since the nullability is defined at type level. Please adapt similar use cases.

**Review by sdeleuze — CHANGES_REQUESTED** [review]
Please refine as suggested for all similar use cases and check with a full build that NullAway does not trigger errors.

File: spring-web/src/main/java/org/springframework/http/ResponseEntity.java (hunk 1 above)
Comment (sbrannen) [reply spring-web/src/main/java/org/springframework/http/ResponseEntity.java]: The `T` here is for a `static` method, which is not directly tied to the `T` in the `ResponseEntity` type declaration -- even though both types are called "T".

So, unless I'm confusing things, I think we need to explicitly declare this as follows.

```java
public static <T extends @Nullable Object> ResponseEntity<T> ok(T body) {
```

Otherwise, the supplied `body` cannot be `null`.

Comment (sdeleuze) [reply spring-web/src/main/java/org/springframework/http/ResponseEntity.java]: Ah good point, so my remark is not relevant.

File: spring-web/src/main/java/org/springframework/http/HttpEntity.java
Comment (ngocnhan-tran1996) [reply spring-web/src/main/java/org/springframework/http/HttpEntity.java]: image

I have removed but the compileJava task build fail

>Within @NullMarked, remember, unannotated types are the same as if they were annotated with @NonNull. Since the bound of E is the same as @NonNull Number, and not @Nullable Number, that means the type argument for E can't be a type that includes null. @Nullable Integer can't be the type argument, then, since that can include null. (In other words: @Nullable Integer is not a subtype of Number.)

https://jspecify.dev/docs/user-guide/#declaring-generics

Comment (ngocnhan-tran1996) [reply spring-web/src/main/java/org/springframework/http/HttpEntity.java]: DITTO

Comment (sdeleuze) [reply spring-web/src/main/java/org/springframework/http/HttpEntity.java]: You need to update the rest of the codebase to fix NullAway errors, with for example here changing the return value to `HttpEntity<? extends @Nullable Object> build()` for example.

File: spring-web/src/main/java/org/springframework/http/ResponseEntity.java [hunk 2]
```diff
@@ -78,7 +78,7 @@
  * @see org.springframework.web.client.RestOperations#getForEntity(URI, Class)
  * @see RequestEntity
  */
-public class ResponseEntity<T> extends HttpEntity<T> {
+public class ResponseEntity<T extends @Nullable Object> extends HttpEntity<T> {
```
Comment (sdeleuze) [inline spring-web/src/main/java/org/springframework/http/ResponseEntity.java:81]: Can you check how `ResponseEntity` of "empty value" translates in a sample application checked with NullAway, using https://github.com/sdeleuze/jspecify-nullaway-demo as a basis maybe?

I guess that would be `ResponseEntity<@Nullable Void>` based on https://github.com/jspecify/jspecify/issues/51 but worth to check.

File: spring-web/src/test/kotlin/org/springframework/http/ResponseEntityKotlinTests.kt [hunk 3]
```diff
@@ -43,4 +43,21 @@ class ResponseEntityKotlinTests {
 		assertThat(responseEntity.body).isNull()
 	}
 
+
+	@Test
+	fun ofNullNullableType() {
```
Comment (sdeleuze) [inline spring-web/src/test/kotlin/org/springframework/http/ResponseEntityKotlinTests.kt:39]: Please add a test for empty value, ie. how `ResponseEntity<@Nullable Void>` translates to Kotlin. Is it `ResponseEntity<Unit>`?

**Review by sdeleuze — COMMENTED** [review]
After discussing with the team, we would like to clarify how entity of empty value translates on both Java and Kotlin signature level before deciding if we move forward.

File: spring-web/src/main/java/org/springframework/http/ResponseEntity.java [hunk 4]
```diff
@@ -117,7 +119,7 @@ public ResponseEntity(HttpHeaders headers, HttpStatusCode status) {
 	 * @param rawStatus the status code value
 	 * @since 7.0
 	 */
-	public ResponseEntity(@Nullable T body, @Nullable HttpHeaders headers, int rawStatus) {
+	public ResponseEntity(T body, @Nullable HttpHeaders headers, int rawStatus) {
```
Comment (PaulNgo-BlueOC) [inline spring-web/src/main/java/org/springframework/http/ResponseEntity.java:122]: Why do we remove the annotation here? Does it mean the T body is not null?

Comment (sdeleuze) [reply spring-web/src/main/java/org/springframework/http/ResponseEntity.java:122]: We remove this annotation there because with this PR, the nullability of `T` with be defined at type level and does not needs to be overridden at method level.

Comment (PaulNgo-BlueOC) [reply spring-web/src/main/java/org/springframework/http/ResponseEntity.java:122]: Please correct me if i'm wrong
IMO, these are 2 completely different stories
- `public class HttptEntity<T extends @Nullable Object>` indicates that the `final T body` is nullable (we just move `@Nullable` from `final @Nullable T body` to the `<....>`
-The `@Nullable T body` in  `public ResponseEntity(@Nullable T body, @Nullable HttpHeaders headers, int rawStatus)`  indicates that the argument passed to the constructor is nullable. `final T body` is nullable vs ...passed to the constructor is nullable.
I would like to reopen the comment to get your feedback. Thanks

Comment (PaulNgo-BlueOC) [inline spring-web/src/main/java/org/springframework/http/ResponseEntity.java:122]: Please correct me if i'm wrong
IMO, these are 2 completely different stories

    public class HttptEntity<T extends @Nullable Object> indicates that the final T body is nullable (we just move @Nullable from final @Nullable T body to the <....>
    -The @Nullable T body in public ResponseEntity(@Nullable T body, @Nullable HttpHeaders headers, int rawStatus) indicates that the argument passed to the constructor is nullable. final T body is nullable vs ...passed to the constructor is nullable.
